DDlog API refactoring Part 1.

ryzhyk opened the pull request:
This is the first step towards making the DDlog API generic so that we
can have multiple implementations, including one that talks to a DDlog
server over the network (#794).

In this commit:

- Split `trait DDlog` into `trait DDlogUntyped` and `trait DDlogTyped`.
  The former works with data represented as records, the latter
  exports methods that work with `DDValue`'s.
- Move a bunch of methods that were in `HDDlog`, but not in `trait DDlog`
  to either one of the above traits or into one of several new traits:
  `DDlogProfiling`, `DDlogDump`, `DDlogInventory`.  This way the entire
  DDlog API is now abstracted away.
- Change method signatures to eliminate any generics.  This way we will
  be able to implement dynamic dispatch for the DDlog API (i.e., pass
  references to a DDlog program as `&dyn DDlogXXX`) in the future.

Kixiron commented on rust/template/src/api/c_api.rs:
```diff
@@ -35,45 +35,88 @@ use std::os::unix::io::{FromRawFd, IntoRawFd, RawFd};
         return libc::size_t::max_value();
     }
 
+    let prog = Arc::from_raw(prog);
```
`ManuallyDrop::new(Arc::from_raw(..))` is less error prone and more clear than calling from_raw and later into_raw

Kixiron commented on rust/template/src/api/c_api.rs:
```diff
@@ -35,45 +35,88 @@ use std::os::unix::io::{FromRawFd, IntoRawFd, RawFd};
+        return ptr::null();
+    }
+
+    let prog = Arc::from_raw(prog);
```
Same as above

Kixiron commented on rust/template/src/api/c_api.rs:
```diff
@@ -35,45 +35,88 @@ use std::os::unix::io::{FromRawFd, IntoRawFd, RawFd};
+        return libc::size_t::max_value();
+    }
+
+    let prog = Arc::from_raw(prog);
```
Arcs again

Kixiron commented on rust/template/src/api/c_api.rs:
```diff
@@ -35,45 +35,88 @@ use std::os::unix::io::{FromRawFd, IntoRawFd, RawFd};
-pub unsafe extern "C" fn ddlog_get_table_id(tname: *const raw::c_char) -> libc::size_t {
-    if tname.is_null() {
+pub unsafe extern "C" fn ddlog_get_table_id(
+    prog: *const HDDlog,
```
To be truly FFI safe while still allowing consumers to take advantage of ddlog we probably want the type used by consumers to be an `Arc<Box<dyn DDlog>>` or similar so that any ddlog runtime can be used while keeping it thread safe since dynamically dispatched objects use fat pointers which aren't ffi safe

Kixiron commented on rust/template/src/api/c_api.rs:
```diff
@@ -35,45 +35,88 @@ use std::os::unix::io::{FromRawFd, IntoRawFd, RawFd};
-pub unsafe extern "C" fn ddlog_get_table_id(tname: *const raw::c_char) -> libc::size_t {
-    if tname.is_null() {
+pub unsafe extern "C" fn ddlog_get_table_id(
+    prog: *const HDDlog,
```
If you want to be clear about the contract that functions expect, you can use `NonNull<HDDlog>` to signify non-null pointers and `Option<NonNull<T>>` for nullable ones. This can be done in both parameter and return positions and allows the backing rust code niceties like `?`

Kixiron replied:
It's undefined behavior, same as what happens with the current code. If nulls are expected (which I don't see why why would be) then `Option<NonNull<T>>` does the same thing with the ability to take a null value as well

ryzhyk commented:
@Kixiron , thanks for the review. I will incorporate your suggestions in the next commit, which will switch the C API over to use dynamic dispatch.

Regarding `Arc`'s, I actually think that we can get rid of `from_raw` altogether.  We should just cast `*HDDlog` to `&HDDlog` (using `&*`) and work with that, since most APIs don't touch the reference count.  I guess I just did not fully understand how `into_raw` works for `Arc`s when writing that code.

Kixiron commented:
Sounds good to me!

ryzhyk commented:
@convolvatron
@mbudiu-vmw 
@blp

blp commented on rust/template/differential_datalog/src/ddlog.rs:
```diff
@@ -60,34 +93,33 @@ pub trait DDlog: Debug {
-    where
-        V: Deref<Target = UpdCmd>,
-        I: Iterator<Item = V>;
+    fn apply_updates_untyped(&self, upds: &mut dyn Iterator<Item = UpdCmd>) -> Result<(), String>;
```
Dynamic?

blp commented on rust/template/differential_datalog/src/replay.rs:
```diff
@@ -48,254 +54,339 @@ where
+/// forwards it to the next handler in the chain.
+///
+/// Error handling:
+/// If the next handler is specified, we invoke it, whether recording succeeded
```
I can't spot a missing verb.

ryzhyk commented:
I will close this PR and re-open it from another branch. I incorporated most feedback.  Most importantly, I renamed

- `trait DDlogUntyped` -> `trait DDlogDynamic`.
- `trait DDlogTyped` -> `trait DDlog` (this trait extends `DDlogDynamic`).
